lib/eco/treelexer/lexer.py

```diff
@@ -80,24 +88,54 @@ def match_one(self, pattern):
         self.exactmatch = False
         return False
 
-    def match_star(self, pattern):
-        while self._inner_match(pattern.c):
-             pass
-        return True
+    def match_star(self, pattern, l=None):
+        if l is None or len(l) == 1:
+            while self._inner_match(pattern.c):
+                pass
+            return True
+
+        if pattern.ng: # non-greedy
+            while True:
+                tmp = self.save_state()
+                if self.match_list(l[1:]):
+                    return True
+                self.load_state(tmp)
+                if self._inner_match(pattern.c):
+                    continue
+                return False
+        else: # greedy
+            tmp = self.save_state()
```

Comment:
I might be wrong, but I don't see how this backtracks more than one level? I was expecting a loop here. Can it deal with cases like `a*aa`?

Reply:
It does, see new commit.

Reply:
`match_list` is recursive and will call `match_star` again, which makes the backtracking work. I tried to do this with a normal loop but failed.